packages/zk-certificates/contracts/AirdropGateway.sol

```diff
@@ -0,0 +1,137 @@
+        require(registrationStartTime < registrationEndTime, "invalid registration time");
+        require(registrationEndTime < claimStartTime, "claim can only start after registration ends");
+        require(claimStartTime < claimEndTime, "invalid claim time");
+        distributions[distributionIndexCounter] = AirdropDistribution(requiredSBTs, registrationStartTime, registrationEndTime, claimStartTime, claimEndTime, msg.sender, tokenAddress, 0, 0, 0, 0);
+        emit DistributionCreated(distributionIndexCounter, msg.sender);
+        distributionIndexCounter++;
+    }
+
+    function getRequiredSBTs(uint distributionId) external view returns (address[] memory) {
+        return distributions[distributionId].requiredSBTs;
+    }
+
+    /* deposit tokens to the distribution
+     we choose to not sending the tokens to distribution contract address directly
+     to just in case distinguish between two distributions with the same token address
+    */
+    function deposit(uint distributionId, uint amount) external onlyRole(CLIENT_ROLE) {
+        IERC20(distributions[distributionId].tokenAddress).transferFrom(msg.sender, address(this), amount);
+        require(distributions[distributionId].claimStartTime > block.timestamp, "claim has already started");
+        require(distributions[distributionId].clientAddress == msg.sender, "only client can deposit");
+        distributions[distributionId].distributionAmount += amount;
+    }
+
+    function withdrawRemainingToken(uint distributionId) external onlyRole(CLIENT_ROLE) {
+        require(distributions[distributionId].clientAddress == msg.sender, "only client can withdraw");
+        require(distributions[distributionId].claimEndTime < block.timestamp, "claim has not ended yet");
+        uint256 amountLeft = distributions[distributionId].distributionAmount - distributions[distributionId].amountClaimed;
+        IERC20(distributions[distributionId].tokenAddress).transfer(msg.sender, amountLeft);
+    }
+    //
+    function register(
+        uint distributionId,
+        uint[2] memory a,
+        uint[2][2] memory b,
+        uint[2] memory c,
+        uint[] memory input) external {
+
+        bytes32 humanID = bytes32(input[verifierWrapper.INDEX_HUMAN_ID()]);
+        uint dAppAddress = input[verifierWrapper.INDEX_DAPP_ID()];
+
+        // check that the public dAppAddress is correct
+        require(
+            dAppAddress == uint(uint160(address(this))),
+            "incorrect dAppAddress"
+        );
+
+        // check the zk proof
+        require(verifierWrapper.verifyProof(a, b, c, input), "invalid proof");
+        require(registeredHumanID[distributionId][humanID] == false, "user has already registered");
+        require(distributions[distributionId].registrationStartTime < block.timestamp, "registration has not started yet");
+        require(distributions[distributionId].registrationEndTime > block.timestamp, "registration has ended");
```

Comment:
For clarity I would add a check that the distribution has already been created. I think this line does it already, but it is not obvious.

Reply:
I added a check for tokenAddress existence to make sure this is the case.